### ISSUE TITLE
Not able to auto add completions to local script

### ISSUE THREAD

**Issue by mrmeszaros** [direnv/direnv#773]
**Describe the bug**
I have a local utility foo, and want to have completions for it.
When I add the completions line to the `.envrc`, it does nothing with bash, and shows and error with fish.

**To Reproduce**
Have a simple script (eg. with python and click) - `foo` with execution rights (`chmod +x foo`):
```python
#!/usr/bin/env python
import click

@click.command()
@click.option('repeat','-r', '--repeat', type=click.IntRange(1, 5), help='Repeat it N times', default=1)
def main(repeat):
  """
  Say Hello a few times.
  """
  for _ in range(repeat):
    print('Hello')

if __name__ == '__main__':
	main()
```
Have a simple `.envrc` alongside:
```
PATH_add .
eval (env _FOO_COMPLETE=source_fish foo)
```
After allowing direnv to load it, I get the following error for fish:
```
direnv: loading ~/Projects/hacking-ground/.envrc
./.envrc:5: syntax error near unexpected token `env'
./.envrc:5: `eval (env _FOO_COMPLETE=source_fish foo)'
direnv: export ~PATH


```
With **bash** (`eval "$(env _FOO_COMPLETE=source_bash foo)"`) there is no error, however, the completions do not get added.

**Expected behavior**
The completions to get added to the shell when running the aforementioned `.envrc` - either with bash or fish.

**Environment**
 - OS: Ubuntu 20.04
 - Shell: fish, bash
 - Direnv version: 2.21.2

**Additional context**
Running the eval lines in the shell adds the completions to the shell.
As I saw in fish, it uses the `complete` shell built-in, in bash, it defines the `_foo_completion` and` _foo_completionup` function, then runs the latter.
Maybe direnv sourcing the output instead of evaling might work, haven't checked every nook and cranny for a workaround.

**Comment by mrmeszaros** [comment]
Did a little bit of digging.
It seems to me, that the `eval` line in bash defines the functions, but it does not get exported back into the main shell.
```bash
PATH_add .
echo '===1==='
typeset -F | grep 'foo'

eval "$(_FOO_COMPLETE=source_bash foo)"
echo '===2==='
typeset -F | grep 'foo'

export -f _foo_completion
export -f _foo_completionetup
echo '===3==='
typeset -F | grep 'foo'
```
Which outputs:
```
direnv: loading ~/Projects/hacking-ground/.envrc
===1===
===2===
declare -f _foo_completion
declare -f _foo_completionetup
===3===
declare -fx _foo_completion
declare -fx _foo_completionetup
direnv: export ~PATH
```
But I think that export works for sub-shells and not for the parent.

You have any idea how could it be implemented? Maybe a new function to the `stdlib`?
Also, other shells would need some extra care - mainly because the `.envrc` is in bash as I saw from the code.

**Comment by zimbatm** [comment]
that's right. The .envrc is evaluated in a bash sub-shell and only environment variables are re-exported to the target shell. This is the only way to make direnv cross-shell compatible. Unfortunately, this means that aliases, functions and other niceties are not supported.

**Comment by mrmeszaros** [comment]
@zimbatm Thanks for clarification!

I strongly believe in the Unix tool philosophy - do only one thing but do it right - so this might marginally step over the boundaries.

Do You think some kind of plugin system could work? Like adding / making place for an extra lib for these niceities. Or just to have a shell-specific source command?

I imagine something like:
```sh
# .envrc
PATH_add .
# adds ./foo to the path

# This command could be loaded from ~/.config/direnv/lib/fishy-business.xxx 
source_fish ./foo-complete.fish
```
I could move `foo-complete.fish` to `~/.config/fish/completions/foo.fish`, and it would work.
But my gut feeling tells me that I should keep related stuff together.
I am not a shell guru, so this might be nigh impossible. Or just not worth the effort.

**Comment by zimbatm** [comment]
> I strongly believe in the Unix tool philosophy

Agreed! the stdlib is already too wide for my taste so I added `source_url` for people to extend direnv.

In the case of per-project completions, I think the right thing to do is to extend $XDG_DATA_DIRS.

Whenever the user tries to complete an executable and that completion isn't installed, the shell should look into something like $dir/completions/$SHELL/$PROGRAM.sh for each folder listed in XDG_DATA_DIRS.
This is something the bash-extras packages is doing already.

**Comment by mrmeszaros** [comment]
@zimbatm Yeah, I might be just wishing for the moon. I am pretty lazy, and want the machine to figure out what I thought, not what I said :smile: 

Anyway, as for the fish autocompletion, I can just add `eval "$(_FOO_COMPLETE=source_bash foo)"` to my `.bashrc` or run  `_FOO_COMPLETE=source_fish foo > ~/.config/fish/completions/foo-complete.fish` for fish.

A common format to define the completions would be nice (that individual shells can interpret and use for their fancy ways) but that might be over complicated - or cannot be generalized enough to have tracktion :thinking:

**Comment by sytherax** [comment]
<img width="1486" height="836" alt="Image" src="https://github.com/user-attachments/assets/c4d337fd-7d24-4229-8ee1-69855f45981f" />

What is the go to solution/workaround for this now?

**Comment by richard-scott** [comment]
Strangely enough, I found this GitHub issue during my searches, as I have this working on Ubuntu 22, but the same direnv config doesn't work on Ubuntu 24.

All I have in my working .envrc file is:
```
# Load bash completion for deploy.sh if available
if [ -f "${PWD}/bin/bash-completions/deploy.sh" ]; then
    source "${PWD}/bin/bash-completions/deploy.sh" 2>/dev/null || true
fi

# Add local bin directory to PATH if it exists
if [ -d "${PWD}/bin" ]; then
    export PATH="${PWD}/bin:${PATH}"
fi
```

I'm trying to figure out why that doesn't work on Ubuntu 24, and now it seems I should be trying to figure out why it **_DOES_** work on Ubuntu 22!

**Comment by richard-scott** [comment]
This is my fix for a bash shell: add the following to/etc/profile.d/direnv.sh

```
# direnv setup - system-wide configuration
# This file sets up direnv hook and enables sourcing files from .envrc in parent shell
#
# The challenge: direnvrc is sourced BEFORE .envrc runs, so it can't see
# variables set by .envrc at the time it's sourced.
#
# Solution: Use PROMPT_COMMAND to check for DIRENV_SOURCE_FILES after
# direnv exports variables. The hook function runs before each prompt,
# so we can check for the variable then.

# Note: The direnv_source_file() helper function is defined in ~/.config/direnv/direnvrc
# which is sourced by direnv before executing .envrc files

if [ -n "$BASH_VERSION" ] && [ -n "$(which direnv 2>/dev/null)" ]; then
  # Set up direnv hook
  eval "$(direnv hook bash)"

  # Function to source files in parent shell
  # Supports colon-separated list of file paths
  source_files() {
      local files_to_source="$1"
      if [ -z "$files_to_source" ]; then
          return 0
      fi

      # Split by colon and source each file
      IFS=':' read -ra FILES <<< "$files_to_source"
      for file in "${FILES[@]}"; do
          # Trim whitespace
          file=$(echo "$file" | xargs)
          if [ -n "$file" ] && [ -f "$file" ]; then
              source "$file"
          fi
      done
  }

  # Hook into PROMPT_COMMAND to check for DIRENV_SOURCE_FILES after direnv exports
  _direnv_source_hook() {
      if [ -n "$DIRENV_SOURCE_FILES" ]; then
          source_files "$DIRENV_SOURCE_FILES"
          # Clear the variable so we don't keep checking
          unset DIRENV_SOURCE_FILES
      fi
  }

  # Add our hook to PROMPT_COMMAND AFTER direnv's hook
  # direnv's hook will be in PROMPT_COMMAND, we just append ours
  # This way direnv works normally, and we just add our functionality
  if [[ "${PROMPT_COMMAND:-}" != *"_direnv_source_hook"* ]]; then
      # If PROMPT_COMMAND already has _direnv_hook, add ours after it
      if [[ "${PROMPT_COMMAND:-}" == *"_direnv_hook"* ]]; then
          PROMPT_COMMAND="${PROMPT_COMMAND};_direnv_source_hook"
      else
          # Otherwise, direnv hook will be added later, so we'll add ours then
          PROMPT_COMMAND="_direnv_source_hook${PROMPT_COMMAND:+;$PROMPT_COMMAND}"
      fi
  fi
fi

```

I as yet have no idea why I do not need ANY of this on Ubuntu 22!
